lightning/src/routing/router.rs

```diff
@@ -349,63 +394,170 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 						// bit lazy here. In the future, we should pull them out via our
 						// ChannelManager, but there's no reason to waste the space until we
 						// need them.
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
+						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, None::<u64>, features.to_context(), 0);
 					}
 				}
 				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
 				// really sucks, cause we're gonna need that eventually.
-				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
+				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, None::<u64>, ChannelFeatures::empty(), 0);
 			}
 		}
 	}
 
-	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
-		if pubkey == *our_node_id {
-			let mut res = vec!(dist.remove(&our_node_id).unwrap().3);
-			loop {
-				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
-					res.last_mut().unwrap().node_features = features.to_context();
-				} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+	let mut payment_paths = Vec::new();
+	let mut tries_left = 10;
+	let mut found_amount_msat = 0;
+	// For every MPP start with the same pre-filled data.
+	let prefilled_targets = targets.clone();
+	let prefilled_dist = dist.clone();
+
+	'route_finding: loop {
+		targets = prefilled_targets.clone();
+		dist = prefilled_dist.clone();
+		'path_finding: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
+			if pubkey == *our_node_id {
+				// TODO: currently pretending that always have blockchain/amount data for channel capacities.
+				// Should handle None better.
+				let mut new_entry = dist.remove(&our_node_id).unwrap();
+				let mut res = vec!(new_entry.3.clone());
```

Comment:
Heh, ok, maybe should have happened before now, but probably worth making the entries here a struct instead of a 6-tuple.